### ISSUE TITLE
dgrid tree + dgrid editor + more than 25 rows → broken rows order

### ISSUE THREAD

**Issue by xdghcnt** [SitePen/dgrid#1130]
1) Going to dgrid/test/data/hierarchicalCountryData.js
2) Adding 25 more Cairo's (https://gist.github.com/xdghcnt/f9611856b3b5923a592a)
3) Going to dgrid/test/tree_editor.html
4) Checking Cairo07
5) Clicking save
→ row order is broken.
Same with `autoSave: true`

dgrid 0.4.0

**Comment by tsofist** [comment]
+1 +1 +1
help help help!

**Comment by jhrabows** [comment]
Few more details and a comment

1. Appears to be specifc to Chrome (was not able to reproduce in other browsers)

2. It can be reproduced with any list of mentioned size. I attached screen shot from Tree_extra_filter.html where additional items were added to the JSON and sort() was added to getChildren():
==
		getChildren: function(parent){
			if(parent.contains){
				// Call filter from the original store to search all objects.
				return store.filter({ type: parent.contains }).sort('name');
			}
		},

==

3. The list appears to be 'almost' sorted. There are only 2 items that are out-of-order: The 1st one and the item that occupies the slot where the 1st item should have been which is in the middle of the list. That second item happens to be the lexicographically 2nd item. In the attached screen shot these items are 'Red Sock' and 'Blue01 Sock' 

Finally a comment: This issue is 4 month old and it has not been triage'd yet. Does Site Pan not realize its seriousness? Almost all grid applications need sorting and a substantial number of these require a tree-structure. This issue prevents dgrid to be used with such applications.

![sortbug](https://cloud.githubusercontent.com/assets/615019/9868448/a80fff7e-5b46-11e5-9e16-c6e509f92737.png)

**Comment by dylans** [comment]
@jhrabows SitePen is a relatively small organization and works on building great things like dgrid as time permits, reinvesting our profits into open source software. In general, if something is urgent, there are ways to let us know, so that bugs don't get lost with other bugs.

In general, our order of priority for dgrid is:
1. Something that is blocking our ability to use dgrid on a customer engagement
2. A support customer is blocked
3. A bug that is clearly blocking key behavior
4. Something we want to fix, to make dgrid better (features, bug fixes, etc.)

A pull request can move something up that priority list, as it gives us a quick start, and we appreciate the help from the community. As noted, we also offer paid support when someone has an urgent issue that they need fixed right away ( http://sitepen.com/support ), and this also helps fund further efforts on open source work. It's not essential to do this, but it is a way to give back and help sponsor further open source efforts.

Regardless, we will take a look at this soon and hopefully have a quick resolution. Just keep in mind, we create dgrid because we think that having a grid component is essential, but we may not be able to drop everything each time a bug is reported. As much as we would love to, resources are limited and we do the best we can. Thanks for understanding!

**Comment by jhrabows** [comment]
Hi dylans,
I do not think Fidelity Investments is a support customer but I guess tree-sorting could fall under item 3 above. I am not yet familiar enough with dgrid code to produce a pull request. However it might be a good idea to create a test for dgrid with few extra rows to demonstrate the behavior.

**Comment by kfranqueiro** [comment]
@jhrabows, thanks for the additional information, and sorry for taking so long to get to this.  This issue was definitely on our radar to get to for the next release.

With our efforts focused on making progress on Dojo 2, we haven't had much time to put into dgrid, but this month we are starting to get a couple of pairs of eyes back on it.  Believe me, I don't like leaving issues like this unattended either.

**Comment by kfranqueiro** [comment]
Oops, I forgot to mention that this issue is specific to cases where no `sort` is specified on the grid, so in the interim you can also work around it by specifying a field to sort by.

While dgrid has logic to avoid this issue, it is ultimately caused by https://github.com/SitePen/dstore/issues/159.

Credit goes to @msssk for the initial diagnosis of this issue earlier this week.

**Comment by tsofist** [comment]
So a year has passed.
The problem remains to this day, but the issue is already closed.
How to solve this problem?

**Comment by kfranqueiro** [comment]
What version of dgrid are you using?  The underlying issue is open in dstore, but it was worked around in dgrid in a changeset that is included in 0.4.1 and 1.0.0.  Meanwhile, I also mentioned a workaround (set `sort`) in my last comment.

**Comment by tsofist** [comment]
@kfranqueiro, thanks for your reply.
Unfortunately, the problem is reproduced on any version dgrid.
Even if sorting is enabled. Eg:
```js
window.grid = new StandardGrid({
  //...
  sort: "name"
}, "treeGrid");
```
Animation problem:
![issue-1130](https://cloud.githubusercontent.com/assets/6843713/15069299/5dc9ede2-1385-11e6-9bed-ee08b5469edc.gif)
